src/hotspot/share/cds/archiveHeapLoader.cpp

```diff
@@ -136,9 +160,39 @@ class PatchUncompressedEmbeddedPointers: public BitMapClosure {
   }
 };
 
+void ArchiveHeapLoader::patch_compressed_pointers(BitMapView bm,
+                                                  FileMapInfo* info,
+                                                  FileMapRegion* map_region,
+                                                  MemRegion region) {
+  // Optimization: if dumptime shift is the same as runtime shift, we can perform a
+  // quick conversion from "dumptime narrowOop" -> "runtime narrowOop".
+  if (_narrow_oop_shift == CompressedOops::shift()) {
+    narrowOop dt_encoded_bottom = info->encoded_heap_region_dumptime_address(map_region);
+    narrowOop rt_encoded_bottom = CompressedOops::encode_not_null(cast_to_oop(region.start()));
+    uint32_t quick_delta = (uint32_t)rt_encoded_bottom - (uint32_t)dt_encoded_bottom;
+
+    log_info(cds)("patching heap embedded pointers: narrowOop 0x%8x -> 0x%8x",
+                  (uint)dt_encoded_bottom, (uint)rt_encoded_bottom);
+    log_info(cds)("CDS heap data relocation quick delta = 0x%x", quick_delta);
+
+    if (quick_delta == 0) {
+      log_info(cds)("CDS heap data relocation unnecessary, quick_delta = 0");
+    } else {
+      PatchCompressedEmbeddedPointersQuick patcher((narrowOop*)region.start(), quick_delta);
+      bm.iterate(&patcher);
+    }
+  } else {
+    log_info(cds)("CDS heap data relocation not possible");
```

Comment:
This message can be confusing because we can do the relocation, just that it wouldn't be as simple as adding a delta.